File: Res/Sorting/Heap Sort/heapSort.dart
void heapsort(List<int> list) {
  int n = list.length;

  // Build a max heap
  for (int i = n ~/ 2 - 1; i >= 0; i--) {
    heapify(list, n, i);
  }
  // Extract elements from heap one by one
  for (int i = n - 1; i > 0; i--) {
    // Move current root (largest element) to the end
    int temp = list[0];
    list[0] = list[i];
    list[i] = temp;

    // Call max heapify on the reduced heap
    heapify(list, i, 0);
  }
}

void heapify(List<int> list, int n, int i) {
  int largest = i; // Initialize largest as root
  int left = 2 * i + 1; // Left child index
  int right = 2 * i + 2;

  // If left child i larger then root
  if (left < n && list[left] > list[largest]) {
    largest = left;
  }
  // If right child is larger than largest so far
  if (right < n && list[right] > list[largest]) {
    largest = right;
  }
  // If largest is not root
  if (largest != i) {
    int swap = list[i];
    list[i] = list[largest];
    list[largest] = swap;

    // Recursively heapify the affected sub-tree
    heapify(list, n, largest);
  }
}

void main() {
  List<int> list = [1, 3, 4, 5, 10, 91, -5, 6];
  heapsort(list);
  print(list);
}

File: Res/Sorting/Heap Sort/heapSort.py
def heapSort(arr):
    n = len(arr)
    mid = n // 2
    for i in range(mid - 1, -1, -1):
        heapify(arr, n , i)
    
    for i in range(n - 1, 0 , -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)
        
def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2
    
    if left < n and arr[left] > arr[largest]:
        largest = left
        
    if right < n and arr[right] > arr[largest]:
        largest = right
    
    if(largest != i):
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

e = [85,96,74,10,20,54,94,36,40,1]
heapSort(e)
print(e)